Docs for custom prefix in Amplify Flutter Storage (#4593)
aws-amplify/docs: src/fragments/lib/storage/flutter/configureaccess/50_customization.mdx

  @override
  Future<String> resolvePrefix({
    required StorageAccessLevel storageAccessLevel,
    String? identityId,
  }) async {
    switch (storageAccessLevel) {
      case StorageAccessLevel.guest:
        return 'myPublicPrefix/';
      case StorageAccessLevel.protected:
        return 'myProtectedPrefix/${identityId!}/';
      case StorageAccessLevel.private:
        return 'myPrivatePrefix/${identityId!}/';
    }
  }
}
```

Then configure the storage plugin with the resolver.

```dart
final storagePlugin = AmplifyStorageS3(prefixResolver: MyCustomPrefixResolver());
await Amplify.addPlugin(storagePlugin);
```

Add the IAM policy that corresponds with the prefixes defined above to enable read, write and delete operation for all the objects under path _myPublicPrefix/_:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::your-s3-bucket/myPublicPrefix/*"
            ],
            "Effect": "Allow"
        }
    ]
}

> 

Reviewer: Missing a code fence

If you want to have custom _private_ path prefix like `myPrivatePrefix/`, you need to add permissions to the IAM policy of your authenticated users. For example, for the given IAM principal (default authenticated role for Amplify projects):

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "cognito-identity.amazonaws.com"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {
                    "cognito-identity.amazonaws.com:aud": "<identity pool ID>"
                },
                "ForAnyValue:StringLike": {
                    "cognito-identity.amazonaws.com:amr": "authenticated"
                }
            }
        }
    ]

> }

Reviewer: Missing code fence

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::your-s3-bucket/myPrivatePrefix/${cognito-identity.amazonaws.com:sub}/*"
            ],
            "Effect": "Allow"
        }
    ]

> }

Reviewer: Missing code fence

## Customization

### Customize Object Key Path

> You can customize your key path by defining a prefix resolver:

Reviewer 2: @fjnoyp - can you please elaborate on the use case here? Why would a customer want to do this basically.

Author: I copied this from the Android and IOS template so if we change this we should probably change Android and iOS docs as well for consistency purposes.

The brevity here is because the previous section of storage explain how private/public/protected uploads are stored with certain prefixes. 

https://docs.amplify.aws/lib/storage/configureaccess/q/platform/android/#protected-access

It might be a bit redundant then to give motivation.  

I wrote a new prompt that elaborates more, but after reading through the entire article I agree with Android/iOS team approach and avoiding redundancy here. 

```
As mentioned above, configuring an access level in storage upload options will cause your files to be uploaded with a _public/_ _protected/_ or _private/_ prefix.

If you would like your items to be stored with a different prefix, you can customize your key path by defining a prefix resolver:
```

Reviewer 2: yes, I like this a lot! Thank you, Kyle!

Author: Great done!

## Customization

### Customize Object Key Path

As mentioned above, configuring an access level in storage upload options will cause your files to be uploaded with a _public/_ _protected/_ or _private/_ prefix.

If you would like your items to be stored with a different prefix, you can customize your key path by defining a prefix resolver:

```dart
// Define your own prefix resolver, that conforms to `StorageS3PrefixResolver`
class MyCustomPrefixResolver implements StorageS3PrefixResolver {
  @override
  Future<String> resolvePrefix({
    required StorageAccessLevel storageAccessLevel,
    String? identityId,
  }) async {
    switch (storageAccessLevel) {
      case StorageAccessLevel.guest:
        return 'myPublicPrefix/';
      case StorageAccessLevel.protected:
        return 'myProtectedPrefix/${identityId!}/';
      case StorageAccessLevel.private:
        return 'myPrivatePrefix/${identityId!}/';
    }
  }
}
```

Then configure the storage plugin with the resolver.

```dart
final storagePlugin = AmplifyStorageS3(prefixResolver: MyCustomPrefixResolver());
await Amplify.addPlugin(storagePlugin);
```

> You will need to update your permissions to allow access to your custom prefix. Go to s3.console.aws.amazon.com then go to IAM -> Roles.

Reviewer 2: I was reading up about this - I think if you make the change just from the AWS Console and not include that somehow in youyr project configuration for Amplify - the next time you do an Amplify push these policies might be removed. I think there has to be another way to do this either by using overrides or extensibility. Can we ask the CLI team about this please?

Author: @AaronZyLee would you know if that happens?

## Customization

### Customize Object Key Path

You can customize your key path by defining a prefix resolver:

```dart
// Define your own prefix resolver, that conforms to `StorageS3PrefixResolver`
class MyCustomPrefixResolver implements StorageS3PrefixResolver {
  @override

>   Future<String> resolvePrefix(
>       {required StorageAccessLevel storageAccessLevel,
>       String? identityId}) async {

Reviewer: ```suggestion
  Future<String> resolvePrefix({
    required StorageAccessLevel storageAccessLevel,
    String? identityId,
  }) async {
```

    switch (storageAccessLevel) {
      case StorageAccessLevel.guest:

>         {
>           return 'myPublicPrefix/';
>         }

Reviewer: ```suggestion
        return 'myPublicPrefix/';
```

      case StorageAccessLevel.protected:

>         {
>           return 'myProtectedPrefix/' + identityId! + '/';
>         }

Reviewer: ```suggestion
        return 'myProtectedPrefix/${identityId!}/';
```

      case StorageAccessLevel.private:

>         {
>           return 'myPrivatePrefix/' + identityId! + '/';
>         }

Reviewer: ```suggestion
        return 'myPrivatePrefix/${identityId!}/';
```

    }
  }
}
```

Then configure the storage plugin with the resolver.

```dart

> AmplifyStorageS3 storage = new AmplifyStorageS3(prefixResolver: MyCustomPrefixResolver());
> Amplify.addPlugin(storage);

Reviewer: ```suggestion
final storagePlugin = AmplifyStorageS3(prefixResolver: MyCustomPrefixResolver());
await Amplify.addPlugin(storagePlugin);
```

```

Add the IAM policy that corresponds with the prefixes defined above to enable read, write and delete operation for all the objects under path _myPublicPrefix/_:

> ```xml
> "Statement": [
>     {
>         "Effect": "Allow",
>         "Action": [
>             "s3:GetObject",
>             "s3:PutObject",
>             "s3:DeleteObject"
>         ],
>         "Resource": [
>             "arn:aws:s3:::your-s3-bucket/myPublicPrefix/*",
>         ]
>     }
> ]
> ```

Reviewer: ```suggestion
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::your-s3-bucket/myPublicPrefix/*"
            ],
            "Effect": "Allow"
        }
    ]
}
```
```

> If you want to have custom _private_ path prefix like _myPrivatePrefix/_, you need to add it into your IAM policy:

Reviewer: ```suggestion
If you want to have custom _private_ path prefix like `myPrivatePrefix/`, you need to add permissions to the IAM policy of your authenticated users. For example, for the given IAM principal (default authenticated role for Amplify projects):

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "cognito-identity.amazonaws.com"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {
                    "cognito-identity.amazonaws.com:aud": "<identity pool ID>"
                },
                "ForAnyValue:StringLike": {
                    "cognito-identity.amazonaws.com:amr": "authenticated"
                }
            }
        }
    ]
}
```

You would add the following to the policy statement:
```

> ```xml
> "Statement": [
>     {
>         "Effect": "Allow",
>         "Action": [
>             "s3:GetObject",
>             "s3:PutObject",
>             "s3:DeleteObject"
>         ],
>         "Resource": [
>             "arn:aws:s3:::your-s3-bucket/myPrivatePrefix/${cognito-identity.amazonaws.com:sub}/*"
>         ]
>     }
> ]
> ```

Reviewer: ```suggestion
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::your-s3-bucket/myPrivatePrefix/${cognito-identity.amazonaws.com:sub}/*"
            ],
            "Effect": "Allow"
        }
    ]
}
```
```

#### Passthrough PrefixResolver

If you would like no prefix resolution logic, such as performing S3 requests at the root of the bucket, create a prefix resolver that returns an empty string:

```dart
class PassThroughPrefixResolver implements StorageS3PrefixResolver {
  @override

>   Future<String> resolvePrefix(
>       {required StorageAccessLevel storageAccessLevel,
>       String? identityId}) async {

Reviewer: ```suggestion
  Future<String> resolvePrefix({
    required StorageAccessLevel storageAccessLevel,
    String? identityId,
  }) async {
```
